### PR TITLE
[PropertyAccess] writeCollection: $key loop iterator as second parameter of $addMethod method called.

### PR DESCRIPTION
**Pull request by cristoforocervino**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

In method writeCollection, added `$key` (loop index iterator) as second parameter of `{$addMethod}` method called.
This extra information allow to fix a problem in Symfony Form CollectionType that lose CollectionElement Index when adding new elements.

With this, when setting `by_reference` to `false` option in CollectionType it will prevent Form Error to lose its key reference the key element of the collection.
The `addMethod` of the entity will be structured like This.

```
 /**
 * Add blockArray
 *
 * @param PostBlock $blockArray
 * @param int       $index
 *
 * @return $this
 */
public function addBlockArray(PostBlock $blockArray, $index = NULL)
{
    if (NULL !== $index) {
        $this->blockArray[$index] = $blockArray;
    } else {
        $this->blockArray[] = $blockArray;
    }
    $blockArray->setPost($this);
    return $this;
}
```

I'm missing something or there is no strongest reason for not to pass `$key` as second parameter of `{$addMethod}`?
Thanks,
Cristoforo

### CODE REVIEW

**Comment by jvasseur** [comment]
This could be considered as a BC-break. If you have an adder like this: `addElement($element, $someOptionalParameter = 'defaultValue')`, you would receive the key instead of the default value after this patch.

**Comment by StevioStudio** [comment]
Good point. I just updated my pull request.

**Comment by stof** [comment]
No change in the testsuite at all ?

**Comment by nicolas-grekas** [comment]
I guess this qualifies as a new feature, so it should be rebased on 3.4 and the base branch should be changed on github. Adding to milestone 3.4

**Comment by StevioStudio** [comment]
I forked from Branch 3.4 instead of master to write tests.
I spotted some problems.
If `$addMethod` does not use `$key` second arguments, than, this is right:
```$axesBefore = $this->getContainer(array(1 => 'second', 3 => 'fourth', 4 => 'fifth'));
$axesMerged = $this->getContainer(array(1 => 'first', 2 => 'second', 3 => 'third'));
$axesAfter = $this->getContainer(array(1 => 'second', 5 => 'first', 6 => 'third'));```

If `$addMethod` uses `$key` second arguments, I expect this:
```$axesBefore = $this->getContainer(array(1 => 'second', 3 => 'fourth', 4 => 'fifth'));
$axesMerged = $this->getContainer(array(1 => 'first', 2 => 'second', 3 => 'third'));
$axesAfter = $this->getContainer(array(1 => 'first', 2 => 'second', 3 => 'third'));```
But I get this:
```$axesBefore = $this->getContainer(array(1 => 'second', 3 => 'fourth', 4 => 'fifth'));
$axesMerged = $this->getContainer(array(1 => 'first', 2 => 'second', 3 => 'third'));
$axesAfter = $this->getContainer(array(1 => 'first', 3 => 'third'));```
Cause `writeCollection` method is not using the `$addMethod` on "second" value due to this check `!in_array($item, $previousValue, true)` (in src/Symfony/Component/PropertyAccess/PropertyAccessor.php at line 688)
and it overrides the value at index "1" of the array.

I'm working to fix this right now.

**Comment by StevioStudio** [comment]
To be able to solve the original problem of this topic, now PropertyAccessor checks if the ACCESS_ADDER method has a second parameter called "key".
Now tests are ok. I added class PropertyAccessorCollectionTest_CarAdderWithKey to describe my case.

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 1]
```diff
@@ -32,23 +32,95 @@
  */
 class PropertyAccessor implements PropertyAccessorInterface
 {
-    private const VALUE = 0;
```
Comment (xabbuh) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: Can you please rebase on the `3.4` branch and change the target branch accordingly to get rid of these unrelated changes?

Comment (StevioStudio) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: Rebased

**Comment by fabpot** [comment]
@StevioStudio Looks like there is still one unrelated commit.

**Comment by StevioStudio** [comment]
@fabpot Is it ok now? :)

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 2]
```diff
@@ -684,9 +690,13 @@ private function writeCollection($zval, $property, $collection, $addMethod, $rem
...
         }
 
-        foreach ($collection as $item) {
-            if (!$previousValue || !in_array($item, $previousValue, true)) {
-                $zval[self::VALUE]->{$addMethod}($item);
+        foreach ($collection as $key => $item) {
+            if ($secondKeyParameter) {
+                $zval[self::VALUE]->{$addMethod}($item, $key);
+            } else {
+                if (!$previousValue || !in_array($item, $previousValue, true)) {
```
Comment (jakzal) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: Could be simplified to `}  elseif (!$previousValue || !in_array($item, $previousValue, true)) {`

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 3]
```diff
@@ -869,6 +882,34 @@ private function isMethodAccessible(\ReflectionClass $class, $methodName, $param
...
+        if ($class->hasMethod($methodName)) {
+            $method = $class->getMethod($methodName);
+
+            if (null === $atIndex) {
+                return !empty(array_filter($method->getParameters(), function (\ReflectionParameter $parameter) use ($parameterName) {
+                    return $parameter->getName() === $parameterName;
+                }));
+            } else {
+                return $method->getNumberOfParameters() >= $atIndex + 1 &&
+                    $method->getParameters()[$atIndex]->getName() === $parameterName;
```
Comment (jakzal) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: no need to break the line

**Comment by nicolas-grekas** [comment]
@StevioStudio rebase needed, would you be able to do it very soon? we're already in feat freeze...

**Comment by nicolas-grekas** [comment]
@StevioStudio btw if can, please rebase on branch 3.4, and change the target branch also here on github.

**Comment by StevioStudio** [comment]
Thanks! Is it ok now? I hope I did everything right, this is the first time I contribute on GitHub.

**Comment by nicolas-grekas** [comment]
Thanks, you did fine!
I just removed method `hasMethodWithParameterNamed` which was not needed, and pushed again on your fork, to also retrigger the CI.

There is a hardcoded convention in this PR: the second parameter *has* to be named `$key`.
Of course, the PropertyAccess component is all about conventions so that may be fine.
That would require a doc PR at least so that ppl can learn about that. Would you please mind opening one?

In your description, you say that this will fix an issue with CollectionType. Is this PR enough to fix it completely? If not, what's missing and are you up to work on it? If yes, what about adding an integration test that ensures the CollectionType bug is fixed?

**Comment by StevioStudio** [comment]
> I just removed method hasMethodWithParameterNamed which was not needed, and pushed again on your fork, to also retrigger the CI.

Thanks! 

> In your description, you say that this will fix an issue with CollectionType. Is this PR enough to fix it completely? If not, what's missing and are you up to work on it? If yes, what about adding an integration test that ensures the CollectionType bug is fixed?

Yes and no. The problem that drive me to this pull request it's a little bit complicate to write. So, I decided to create a PDF to show you the original problem in order to take suggestions from you and understand if this is the right way to fix it.
[Symfony Form Error Bug.pdf](https://github.com/symfony/symfony/files/1368785/Symfony.Form.Error.Bug.pdf) 

Thank you for your patience and I hope PDF will help.

**Comment by nicolas-grekas** [comment]
@xabbuh @dunglas LGTM here, but not my comfort zone, WDYT?

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 4]
```diff
@@ -76,6 +76,11 @@ class PropertyAccessor implements PropertyAccessorInterface
     /**
      * @internal
      */
+    const ACCESS_ADDER_WITH_KEY = 6;
```
Comment (xabbuh) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: I would move this after the `ACCESS_REMOVER` constant to avoid confusion.

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 5]
```diff
@@ -857,7 +868,7 @@ private function findAdderAndRemover(\ReflectionClass $reflClass, array $singula
      *
      * @return bool Whether the method is public and has $parameters required parameters
      */
-    private function isMethodAccessible(\ReflectionClass $class, $methodName, $parameters)
+    private function isMethodAccessible(\ReflectionClass $class, $methodName, $parameters, \ReflectionMethod &$method = null)
```
Comment (xabbuh) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: Not sure if it's really is a good idea to make the method have this side effect. I think I would rather split it to extract the part that is getting the `ReflectionMethod` instance.

**Comment by xabbuh** [comment]
Generally I wonder, why the key is expected to be the second argument of the adder method.

**Comment by xabbuh** [comment]
Also I do not understand why an adder method should have a key at all. IMO that's a setter method and we already have support for them.

**Comment by StevioStudio** [comment]
@xabbuh I agree. But I did not find another est to fix tue huge problem that I show in my PDF.

**Comment by nicolas-grekas** [comment]
@StevioStudio can you please rebase + retarget for the master? Please also note that since master supports php >=7.1, you can use more language features (eg private const, etc.?)

Status: needs work

**Comment by StevioStudio** [comment]
@nicolas-grekas I first need to know some opinion about the problem I shown in my pdf [Symfony Form Error Bug.pdf](https://github.com/symfony/symfony/files/1368785/Symfony.Form.Error.Bug.pdf).
Because I somehow agree with @xabbuh about the fact that this is not a good change for PropertyAccess.

Maybe it would be better to find a way to fix this SymfonyForm bug without change the PropertyAccess.
What do you think about it?

**Comment by fabpot** [comment]
I've just re-read this thread. From my understanding, everybody agrees that the current PR is not mergeable as is. I also understand by reading the PDF that the original issue is on the Form component. So, I'm going to close this PR, and I suggest that we create an issue with the PDF attached explaining the core issue (maybe @HeahDude can have a look at the form issue and give us some insights).
